elastalert/elastalert.py

```diff
@@ -553,62 +554,70 @@ def load_rule_changes(self):
-                    self.handle_error("Error running rule %s: %s" % (rule['name'], e), {'rule': rule['name']})
-                else:
-                    old_starttime = pretty_ts(rule.get('original_starttime'), rule.get('use_local_time'))
-                    logging.info("Ran %s from %s to %s: %s query hits, %s matches,"
-                                 " %s alerts sent" % (rule['name'], old_starttime, pretty_ts(endtime, rule.get('use_local_time')),
-                                                      self.num_hits, num_matches, self.alerts_sent))
-                    self.alerts_sent = 0
-
-                self.remove_old_events(rule)
-
-            if next_run < datetime.datetime.utcnow():
-                # We were processing for longer than our refresh interval
-                # This can happen if --start was specified with a large time period
-                # or if we are running too slow to process events in real time.
-                logging.warning("Querying from %s to %s took longer than %s!" % (old_starttime, endtime, self.run_every))
-                continue
+        for rule in self.rules:
+            # Set endtime based on the rule's delay
+            delay = rule.get('query_delay')
+            if hasattr(self.args, 'end') and self.args.end:
+                endtime = ts_to_dt(self.args.end)
+            elif delay:
+                endtime = ts_now() - delay
+            else:
+                endtime = ts_now()
+
+            try:
+                num_matches = self.run_rule(rule, endtime, self.starttime)
+            except EAException as e:
+                self.handle_error("Error running rule %s: %s" % (rule['name'], e), {'rule': rule['name']})
+            else:
+                old_starttime = pretty_ts(rule.get('original_starttime'), rule.get('use_local_time'))
+                logging.info("Ran %s from %s to %s: %s query hits, %s matches,"
+                             " %s alerts sent" % (rule['name'], old_starttime, pretty_ts(endtime, rule.get('use_local_time')),
+                                                  self.num_hits, num_matches, self.alerts_sent))
+                self.alerts_sent = 0
+
+            self.remove_old_events(rule)
+
+        if next_run < datetime.datetime.utcnow():
+            # We were processing for longer than our refresh interval
+            # This can happen if --start was specified with a large time period
+            # or if we are running too slow to process events in real time.
+            logging.warning("Querying from %s to %s took longer than %s!" % (old_starttime, endtime, self.run_every))
+            return
+
+        # Only force starttime once
+        self.starttime = None
 
-            # Only force starttime once
-            starttime = None
+        if not self.args.pin_rules:
+            self.load_rule_changes()
 
-            if not self.args.pin_rules:
-                self.load_rule_changes()
+        # Wait before querying again
```

Comment:
I think this should go in the start loop. Sleeping before returning makes no sense if you are calling run_all_rules directly.